README.md
# LuaNet

Lua module for creating multilayer neural networks


<!-- ABOUT THE PROJECT -->
## About The Project
LuaNet, a module for using neural networks in Lua! **This module is a tool easy to use for beginners and at the same time powerful**  enough to be used in small projects
Inspired by the [Toy Neural Network](https://github.com/CodingTrain/Toy-Neural-Network-JS), this module aims to reach all people who are entering the world of machine learning and want to implement it in their projects and games.
In view of this, this module comes with functions for creating neural networks, as well as manipulating matrices. All of this is compatible with [LOVE2D]([https://love2d.org), as well as with all other platforms and projects that use lua.

As mentioned earlier, the objectives of this module are:
* Be friendly and easy to use for beginners
* Be compatible with the maximum of platforms, projects, systems and etc. that use lua

It is worth mentioning that this project **is not a library for high-level projects**, that is: that it has all the capacities and technologies for application in professional and academic environments, for that I recommend using tools such as pytorch.




<!-- GETTING STARTED -->
## Getting Started
Before anything I highly recommend that you look at the documentation in the ``` \Doc``` folder
This is an example of how you may give instructions on setting up your project locally.
To get a local copy up and running follow these simple example steps.

### Prerequisites

Only a single prerequisite:
* lua


### Installation

1. Download and extract to anywhere you wish. **It is important that the files "NeuralLua.lua", "Matrix.lua" and "csvHandler.lua" stay together in the same folder**
2. Sometimes you may have to adjust the location of the files in the code, if any of these files cannot find the other open the code and follow the instructions commented above the "require" of each file. **For a complete view of the error and how to fix it see the documentation,"module issue" section**




<!-- USAGE EXAMPLES -->
## Basic Usage
1 Import the module
```LUA
luanet = require("path".."NeuralLua")
matrix = require("path".."Matrix") --sometime you may need the matrix module
```
2. create the neural network
```LUA
network = luanet.new({2,3,2}) -- creates a neural network with 2 inputs, 1 hidden layer of 3 neurons and 2 outputs
```
3. Trains the neural network
```LUA
network = nn.train(network,{inputx1,inputx2},{outputy1,outputy2})
```
4. Make a prediction
```LUA
pr = nn.preddict(network,{output1,output2})
```
4. The prediction always returns a matrix table so you can convert it to an array or print using the following commands
```LUA
array = matrix.toarray(pr)
--or print
matrix.print(pr)
```

**I highly recommend that you look at the documentation in the ``` \Doc``` folder**



<!-- ROADMAP -->
## Roadmap

 - More activation functions
 - Normalize dataset function
 - Functions for handling images
 - Functions for handling text
 - Portuguese Documentation (Documentação em portugues)

<!-- CONTRIBUTING -->
## Contributing

Pull requests are **very welcome**. For major changes, please open an issue first to discuss what you would like to change.

Please make sure to update tests as appropriate.



<!-- LICENSE -->
## License

Distributed under the MIT License. See `LICENSE` for more information.



<!-- CONTACT -->
## Author

 - João P. Maia 




<!-- ACKNOWLEDGEMENTS -->
## Bibliography
* [Codding train](https://www.youtube.com/channel/UCvjgXvBlbQiydffZU7m1_aw)
* [3Blue1Brown]([https://www.youtube.com/channel/UCYO_jab_esuFRV4b17AJtAw)
* [Toy Neural Network](https://github.com/CodingTrain/Toy-Neural-Network-JS)

I researched and studied many resources for this list. I would like to name these pos was inspired by these channels that I started with this project. For a complete list check the documentation


Documantation/changelog.txt
v1.0.0 - Inital Commit 5/6/2020

v1.1.0 - Bugfixes 19/8/2020
+ [Fixed] Little bug when loading NN table in the activation functions 
+[FIxed] Local and Global Variables
+[Added] check function in NeuralLua.lua file

Lib/Matrix.lua
-- Matrix.lua 
--This file handles the creation and manipulation os the matrixes. 
--Author: Joao P Maia
--
--
local mat = {}
local write = io.write
math.randomseed(os.time())


--
-- Creating matrix
--

--New specifed matrix
function mat.new(rows,cols,values)
  local grid ={}
  grid.rows = rows
  grid.cols = cols
  grid.matrix = {}
  local j = 1
  for i=1,rows do
    grid.matrix[i] = {}
    for k=1,cols do
      grid.matrix[i][k] = values[j]
      j = j+1
    end
  end
    return grid
  end
  --
--New matrix fom 1d array
function mat.newfromarray(array)
  local grid ={}
  grid.rows = #array
  grid.cols = 1
  grid.matrix = {}
  local j = 1
  
  for i=1,grid.rows do
    grid.matrix[i] = {}
    for k=1,grid.cols do
      grid.matrix[i][k] = array[j]
      j = j+1
    end
  end
    return grid
  end
  --
--New Zero matrix
function mat.newzero(rows,cols)
  local grid ={}
  grid.rows = rows
  grid.cols = cols
  grid.matrix = {}
  local j = 1
  for i=1,rows do
    grid.matrix[i] = {}
    for k=1,cols do
      grid.matrix[i][k] = 0
      j = j+1
    end
  end
    return grid
  end
  --
  --New random matrix
  function mat.newrdn(rows,cols,positive)
  positive = false or positive
  local grid ={}
  grid.rows = rows
  grid.cols = cols
  grid.matrix = {}
  local j = 1
  for i=1,rows do
    grid.matrix[i] = {}
    for k=1,cols do
      if positive then
        grid.matrix[i][k] = math.random()
      else
        grid.matrix[i][k] = math.random(-1,0) + math.random()
      end
      j = j+1
    end
  end
    return grid
end
--New matrix from a given label, for classification
function mat.newfromlabel(rows,label_index)
  local grid ={}
  grid.rows = rows
  grid.cols = 1
  grid.matrix = {}
  local j = 1
  for i=1,rows do
    grid.matrix[i] = {}
    for k=1,grid.cols do
      if i == label_index then
        grid.matrix[i][k] = 1
      else
        grid.matrix[i][k] = 0
      end
    end
  end
    return grid
  end
  --
  
--
--  Matrix operations
--
  
-- Inner Sum
  function mat.innersum(m1)
    local result = 0
    for i=1,result.rows do
      for j=1,result.cols do
          result = m1.matrix[i][j] + result
      end
    end
    return result
  end
  
  
  
-- Element sum
  function mat.sum(m1,m2)
    --checking if there is an error
    if(m1.cols ~= m2.cols or m1.rows ~= m2.rows) then
      print("erro in subtraction, cols rows erro")
    end
    --
    local result = mat.newzero(m1.rows,m2.cols)
    for i=1,result.rows do
      for j=1,result.cols do
          result.matrix[i][j] = m1.matrix[i][j] + m2.matrix[i][j]
      end
    end
    return result
  end
  --  
-- Element subtraction
  function mat.sub(m1,m2)
    --checking if there is an error
    if(m1.cols ~= m2.cols or m1.rows ~= m2.rows) then
      print("Erro in matrix subtraction, cols rows erro")
    end
    --
    local result = mat.newzero(m1.rows,m2.cols)
    for i=1,result.rows do
      for j=1,result.cols do
          result.matrix[i][j] = m1.matrix[i][j] - m2.matrix[i][j]
      end
    end
    return result
  end
  --    
-- Element multiply
  function mat.mult(m1,m2)
    local result = mat.newzero(m1.rows,m2.cols)
    for i=1,result.rows do
      for j=1,result.cols do
          result.matrix[i][j] = m1.matrix[i][j] * m2.matrix[i][j]
      end
    end
    return result
  end
  --    
  --Dot product matrix
  function mat.dot(m1,m2)
    if m1.cols ~= m2.rows then
      print("Erro in matrix dot, cols rows erro")
      return
    end
    
    
    local result = mat.newzero(m1.rows,m2.cols)
    for i=1,result.rows do
      for j=1,result.cols do
        local sum = 0
        -- dot
        for k=1,m1.cols do
          sum = sum + m1.matrix[i][k] * m2.matrix[k][j]
        end
        result.matrix[i][j] = sum
      end
    end
    return result
  end
  --
 -- Scalar multiply
  function mat.mult1D(m1,scalar)
    local result = mat.newzero(m1.rows,m1.cols)
    for i=1,result.rows do
      for j=1,result.cols do
          result.matrix[i][j] = m1.matrix[i][j] * scalar
      end
    end
    return result
  end
  --
  
--
--  Matrix modifications
--
  
  --Traspose
  function mat.transpose(m1)
    local result = mat.newzero(m1.cols,m1.rows)
    for i=1,m1.rows do
      for j=1,m1.cols do
        result.matrix[j][i] = m1.matrix[i][j]
      end
    end
    return result
  end
  --
  --Apply a function in each number in the matrix
  function mat.map(m1,func)
    local result = mat.newzero(m1.rows,m1.cols)
    for i=1,result.rows do
      for j=1,result.cols do
        result.matrix[i][j] = func(m1.matrix[i][j])
      end
    end
    return result
  end
  --
  --Matrix to array
  function mat.toarray(m1)
    local result = {}
    for i=1,m1.rows do
      for j=1,m1.cols do
        table.insert(result,m1.matrix[i][j])
      end
    end
    return result
  end
  --
  
--
--  Matrix uncategorized functions
--

function mat.print(m,debug,separator)
  --DEBUG
  debug = debug or 0
  if debug ~= 0 then
    print("Matrix "..m.rows.."X"..m.cols..", ".. m.cols*m.rows.." elements")
  end
  --Separator
  separator = separator or " | "
  
  --Actual function
  for c=1,m.rows do
    local r = 1
    while r <= m.cols do
      write(tostring(m.matrix[c][r]))
      write(separator)
      r = r+1
    end
    r = 1
    print()
  end
end  
  

--End
return mat


Lib/csvHandler.lua
local csv = {}
math.randomseed(os.time())
--AUX
function shuffle(array)
    -- fisher-yates
    local output = { }
    local random = math.random

    for index = 1, #array do
        local offset = index - 1
        local value = array[index]
        local randomIndex = offset*random()
        local flooredIndex = randomIndex - randomIndex%1

        if flooredIndex == offset then
            output[#output + 1] = value
        else
            output[#output + 1] = output[flooredIndex + 1]
            output[flooredIndex + 1] = value
        end
    end

    return output
end

-- read csv
function csv.readfile(arq)
   local file = io.open(arq, "r");
 local arr = {}
 for line in file:lines() do
    table.insert (arr, line);
 end
return arr
end
-- line to array
function csv.parseCSVLine (line,sep) 
	local res = {}
	local pos = 1
	sep = sep or ','
	while true do 
		local c = string.sub(line,pos,pos)
		if (c == "") then break end
		if (c == '"') then
			-- quoted value (ignore separator within)
			local txt = ""
			repeat
				local startp,endp = string.find(line,'^%b""',pos)
				txt = txt..string.sub(line,startp+1,endp-1)
				pos = endp + 1
				c = string.sub(line,pos,pos) 
				if (c == '"') then txt = txt..'"' end 
				-- check first char AFTER quoted string, if it is another
				-- quoted string without separator, then append it
				-- this is the way to "escape" the quote char in a quote. example:
				--   value1,"blub""blip""boing",value3  will result in blub"blip"boing  for the middle
			until (c ~= '"')
			table.insert(res,txt)
			assert(c == sep or c == "")
			pos = pos + 1
		else	
			-- no quotes used, just look for the first separator
			local startp,endp = string.find(line,sep,pos)
			if (startp) then 
				table.insert(res,string.sub(line,pos,startp-1))
				pos = endp + 1
			else
				-- no separator found -> use rest of string and terminate
				table.insert(res,string.sub(line,pos))
				break
			end 
		end
	end
	return res
end

--Export the csv to train
function csv.totrain(arq,sep)
  sep = sep or ","
  --Load the file
  local file = csv.readfile(arq)
  --Parse data
  local data_parse = {}
  for i=2, #file do
    table.insert(data_parse,csv.parseCSVLine(file[i],sep))
  end
  -- shuffle the data
  local data = shuffle(data_parse)
  return data
end
return csv


Demos/Iris/iris.lua
--Import LuaNet module, contains all neural network functions
nn = require('Lib.NeuralLua')

--Import matrix.lua module, takes care of the matrix part
mat = require('Lib.Matrix')

--Create or train
print("Would you like to create a neural network or load the already trained network?? (Create type 0, Load type 1)")
local train = tonumber(io.read())

if train == 0 then
  --Creates a neural network with 4 inputs, a hidden layer of 5 neurons and 3 outputs
  Brain = nn.new({4,5,3})

  print("Press anything to to train")
  io.read()

  --Load the dataset in csv, first argument defines values that must be transformed into int, second defines y and lastly the file location
  dataset = nn.newdataset({5},5,'Demos/Iris/iris_dataset.csv')

  --training options
  options = {lr = 0.01,
     epochs = 150,
     validation = 0.20,
     debug = true,
     label = true }
  --Trains the neural network
  Brain = nn.traindataset(Brain,dataset,options)
  
else
  
  --Load the dataset in csv, first argument defines values that must be transformed into int, second defines y and lastly the file location
  dataset = nn.newdataset({5},5,'Demos/Iris/iris_dataset.csv')
  
  Brain = nn.load('Demos/Iris/Saved/',"NNIris")
  print("Model loaded")
end
--

--preddicting
print("Finished, start preddicting")
print("sepal length: ")
x1 = io.read()
print("sepal width: ")
x2 = io.read()
print("petal length: ")
x3 = io.read()
print("petal width: ")
x4 = io.read()

--Prediction based on inputs
pr = nn.preddict(Brain,{x1,x2,x3,x4})

--Turn prediction into a label
lbl = nn.getlabel(dataset,pr)
print(lbl)


Lib/NeuralLua.lua
--NeuralLua file
--Author: Joao P Maia
--This file takes care of all the neual network part
local nn = {}
local matrix = require('Lib.Matrix') -- MODULE ISSUE||If the "module issue" error described in the documentation has occurred. Please change and path to point to Matrix.lua file
local csv = require('Lib.csvHandler') --MODULE ISSUE ||If the "module issue" error described in the documentation has occurred. Please change and path to point to csvHandler.lua file


--Aux Functions
math.randomseed(os.time())
function labelToArray(dataset_size,item)
  local arr = {}
  for i=1,dataset_size do
    if i == item then
      table.insert(arr,1)
    else
      table.insert(arr,0)
    end
  end
return arr
end
function tabledel(tb,index)
  local tb_new = {}
  for i=1,#tb do
    if i ~= index then
      table.insert(tb_new,tb[i])
    end
  end
  return tb_new
end

function shuffle(array)
    -- fisher-yates
    local output = { }
    local random = math.random

    for index = 1, #array do
        local offset = index - 1
        local value = array[index]
        local randomIndex = offset*random()
        local flooredIndex = randomIndex - randomIndex%1

        if flooredIndex == offset then
            output[#output + 1] = value
        else
            output[#output + 1] = output[flooredIndex + 1]
            output[flooredIndex + 1] = value
        end
    end

    return output
end

function normalize(x,max,min)
  return (x-min)/(max-min) 
end
function IndexOf(t,val)
    for k=1, #t do 
        if tostring(t[k]) == tostring(val) then return k end
    end
end
function contains(table, val)
   for i=1,#table do
      if table[i] == val then 
         return true
      end
   end
   return false
end
function round(number, decimals)
    local decimals = 10 or decimals
    local power = 10^decimals
    return math.floor(number * power) / power
end

--
--Activations functions
--

--Sigmoid
function sigmoid(x)
  return 1/ (1+math.exp(-x))
end
--Sigmoid Derivative
function Dsigmoid(x)

  return x * (1-x)
end
--Hyperbolic Tangent
function tanh(x)
  return math.tanh(x)
end
--Hyperbolic Tangent Derivative
function Dtanh(x)
  return 1-(x*x)
end

--
--Basic Functions
--

function nn.check()
  print("Nueral Netwotk module loaded || NeuralNet V1.1.0 by JP Maia")
    
end

--Create a NN table
function nn.new(neurons,activation)
  local activation = activation or "sigmoid"
  
   local neural = {}
  -- store each neuron numbers
  neural.neurons = neurons
  --Store the weights
  neural.weights = {}
  --Activation and de activation function
  neural.F_activation = activation
  if activation == "sigmoid" then
    neural.activation  = sigmoid
    neural.Dactivation  = Dsigmoid
  else
    neural.activation  = tanh
    neural.Dactivation  = Dtanh
  end
  
  --Generate radom weights
  local i = 1
  while i<#neurons do
    local weight_n = matrix.newrdn(neurons[i+1],neurons[i])
    table.insert(neural.weights,weight_n)
    i = i+1
  end
  
  --Store BIAS
  neural.bias = {}
  i = 1
  while i<#neurons do
    local bias_n = matrix.newzero(neurons[i+1],1)
    table.insert(neural.bias,bias_n)
    i = i+1
  end

  return neural
end
--
--Predicts
function nn.preddict(neural,x,trainning)
  --Transfrom x in matrix
  local input = matrix.newfromarray(x)
  
  -- y = inputs
   local y = {}
   table.insert(y,input)
   
  -- preddicting
  local i = 1
  while i <= #neural.weights do
    Y_n = matrix.dot(neural.weights[i],y[i])
    Y_n = matrix.sum(Y_n,neural.bias[i])
    Y_n  = matrix.map(Y_n,neural.activation)
    Y_n = matrix.map(Y_n,round)
    table.insert(y,Y_n)
    i = i+1
  end
  
  if trainning then
    return y
  end
  
  return y[#y]
  
  end


--
--Trainning
function nn.train(neural,input_x,target_y,lr)
  -- lr = learning rate
  local lr = lr or 0.5
  --preparation
  local y = nn.preddict(neural,input_x,1)
  local target_y = matrix.newfromarray(target_y)
  local errors = {}
  --step A, calcutae overall error
  local total_err = matrix.sub(target_y,y[#y])
  
  table.insert(errors,total_err)
  
  -- calculating erros in each layer
   local i = #neural.weights
   local k = 1
   while i > 1 do
    local layer_trasposed = matrix.transpose(neural.weights[i])
    local erro_n = matrix.dot(layer_trasposed,errors[k])
    table.insert(errors,erro_n)
    i = i-1
    k = k+1
  end
  
  -- calculating the deltas
  local deltas = {}
  i = 1
  k = #y
  local  b = #neural.bias
  while i <= #neural.weights do
  local gradient_n = y[k]
  local gradient_n = matrix.map(gradient_n,neural.Dactivation)
  
  -- mutiplicar gradiente pelo erro
  gradient_n = matrix.mult(gradient_n,errors[i])
  gradient_n = matrix.mult1D(gradient_n,lr)
  
  -- Adiciona o gradiente ao BIAS
  neural.bias[b] = matrix.sum(neural.bias[b],gradient_n)
  b = b-1
  
  -- transpor o que esta saindo
  local exiting = matrix.transpose(y[k-1])
  local delta =  matrix.dot(gradient_n,exiting)
  
  table.insert(deltas,delta)
  
  
  i = i+1
  k = k-1
  end

  -- aplying deltas
  b = #neural.bias
  i = 1 
  k = #neural.weights
  while i <= #neural.weights do
    neural.weights[i] = matrix.sum(neural.weights[i],deltas[k])
    i = i+1
    k = k-1
  end
  
  -- aplying BIAS
  b = #neural.bias
  i = 1 
  k = #neural.weights
  while i <= #neural.weights do
    neural.weights[i] = matrix.sum(neural.weights[i],deltas[k])
    i = i+1
    k = k-1
  end
  
  
  return neural

end

--
--Dataset Trainning
--

--Geral data to trainnig
function nn.newdataset(labels,target,file,sep)
  local sep =  sep or ","
  local target_to_num
  
  local dataset = {}
  dataset.x = {}
  dataset.y = {}
  dataset.Xlabel = {}
  dataset.Ylabel = {}

  local data = csv.totrain(file,sep)
  local teste = csv.totrain(file,sep)
  
  -- Check if the targets need conversion
  for i=1,#labels do
    if labels[i] == target then
      target_to_num = true
      -- new labels table without the target
      labels = tabledel(labels,i)
    end
  end
  
  -- convert label targets to numbers
  if target_to_num then
    -- convert label to number
    for i=1,#data do
      if contains(dataset.Ylabel,data[i][target]) == false then
      table.insert(dataset.Ylabel,data[i][target])
      end
    data[i][target] = IndexOf(dataset.Ylabel,data[i][target])
    end
  end
  

  
  
  -- convert the rest here
  for j=1,#labels do
    dataset.Xlabel[j] = {} -- create a table for each label
    local index = labels[j]
    for i=1,#data do
      if contains(dataset.Xlabel[j],data[i][index]) == false then
      table.insert(dataset.Xlabel[j],data[i][index])
      end
    data[i][index] = IndexOf(dataset.Xlabel[j],data[i][index])
    end
  end
  
  -- convert everything to number
  for c=1,#data do
    for r=1,#data[c] do
      data[c][r] = tonumber(data[c][r])
    end
  end
  
  --pass to labels
  for i=1, # data do
    dataset.y[i] = data[i][target]
    data[i] = tabledel(data[i],target)
  end
  
  -- if target is label pass a arr as value
  for i=1,#dataset.y do
    dataset.y[i] = labelToArray(#dataset.Ylabel,dataset.y[i])
  end
  
  -- pass everything to dataset table
  for i=1,#data do
    dataset.x[i] = data[i]
  end

  return dataset

end
--Train in a dataset
function nn.traindataset(neural,dataset,options)
  --defalut options
  options = options or {}
  local lr = options.lr or 0.1
  local epochs = options.epoch or 150
  local validation = options.validation or 0.20
  local debug = options.debug or true
  local label = options.label or false
  
  --starting
  local trainnig_data = #dataset.x - math.floor(#dataset.x*validation) 
  local validation_data = math.floor(#dataset.x*validation)
  
  
  --trainnig for given epochs
  local run = 0
  while run<epochs do
    if debug then
      print("Epoch "..run..' of '..epochs)
    end
    for t=1,trainnig_data do
      neural = nn.train(neural,dataset.x[t],dataset.y[t],lr)
    end
    run = run+1
  end

  --validate trannig
  --Accuracy
  local acc = 0 
  for i=trainnig_data,#dataset.x do
    local preddiction = nn.preddict(neural,dataset.x[i])
    local true_y = dataset.y[i]
    local true_y = matrix.newfromarray(true_y)
    
    --accuracy for classification
    if label then
      if nn.getlabel(dataset,true_y) == nn.getlabel(dataset,preddiction) then
        acc = acc+1
      end
    end
    
  end
  
  if debug then
    if label then
    print("Done training, "..acc.." correct awnsers of "..(#dataset.x-trainnig_data).." guesses. Estimated acuracy: "..acc/(#dataset.x-trainnig_data))
    end
  end
  return neural
end
--transform a matrix into a label
function nn.getlabel(dataset,mat)
  local arr = matrix.toarray(mat)
  local best = 1
  for i=1,#arr do
    if arr[i] > arr[best] then
      best = i
    end
  end
  return dataset.Ylabel[best]
end

--
--Save and Load functions
--

--Save NN table
function nn.save(neural,path,personal_name)
  path = path or nil
  personal_name = personal_name or math.random(0,1000)
  
  --NN Info
  local name = "NN"..personal_name
  name = tostring(name)
  
  --NN Save geral Info
  local info = io.open(path..name..".txt", "a")
  info:write(name.."\n")
  info:write(#neural.neurons.."\n")
  info:write(#neural.weights.."\n")
  info:write(#neural.bias.."\n")
  info:write(neural.F_activation.."\n")
  for i=1,#neural.neurons do
    info:write(neural.neurons[i].."\n")
  end

  --NN Save geral Neurons
  local n = io.open(path..name.."Neurons"..".txt", "a")
  for k=1,#neural.neurons do
    local neurons = neural.neurons[k]
    n:write(neurons.."\n")
  end
    
    
    
  --NN Save weights
  for k=1,#neural.weights do
    local w = io.open(path..name.."Weights"..k..".txt", "a")
    local weight = neural.weights[k]
    for i=1,weight.rows do
      for j=1,weight.cols do
            w:write(weight.matrix[i][j].."\n")
      end
    end
  end

  
  --Save Bias
  for k=1,#neural.bias do
    local b = io.open(path..name.."Bias"..k..".txt", "a")
    local bias = neural.bias[k]
    for i=1,bias.rows do
      for j=1,bias.cols do
            b:write(bias.matrix[i][j].."\n")
      end
    end
  end
  

end
--Load NN
function nn.load(path,name)
  
  --Create NN table
  local neural = {}
  neural.weights = {}
  neural.bias = {}
  neural.neurons = {}
  
  
  --Get Info
  local inf = io.open(path..name..".txt", "r")
  local info = {}
  for line in inf:lines () do
    table.insert(info,line)
  end 

  name = info[1]
  local neurons_N = info[2]
  local weights = tonumber(info[3])
  local bias = tonumber(info[4])
  neural.F_activation = info[5]
  
  for i = 6, #info do
    table.insert(neural.neurons,tonumber(info[i]))
  end
  
  
  --Get neurons
    local neurons = {}
    local neurons_path = path..name.."Neurons"..".txt"
    local n = io.open(neurons_path, "r")
  for line in n:lines () do
      table.insert(neurons,tonumber(line))
  end

  
  --Get weights
  local weights_values = {}
  for i=1,weights do
    local weights_path = path..name.."Weights"..i..".txt"
    local w = io.open(weights_path, "r")
    weights_values[i] = {}
    for line in w:lines () do
      table.insert(weights_values[i],tonumber(line))
    end 
  end
  
  
  
  --Get BIAS
  local bias_values = {}
  for i=1,bias do
    local bias_path = path..name.."Bias"..i..".txt"
    local b = io.open(bias_path, "r")
    bias_values[i] = {}
    for line in b:lines () do
      table.insert(bias_values[i],tonumber(line))
    end 
  end
  
  
  
  --Weights to neural
  for i=1,#neurons-1 do
    neural.weights[i] = matrix.new(neurons[i+1],neurons[i],weights_values[i])
  end
  
  
  --Bias to neural
  for i=1,#neurons-1 do
    neural.bias[i] = matrix.new(neurons[i+1],1,bias_values[i])
  end
  
  
  
  
  if neural.F_activation == "tanh" then
    neural.activation = tanh
    neural.Dactivation = Dtanh
  else
    neural.activation = sigmoid
    neural.Dactivation = Dsigmoid
  end
  
  return neural
end

--
--Neuroevolution functions
--

--Copy
function nn.copy(neural)
  local new_neural = neural
  return new_neural
end
--Mutate
function nn.mutate(neural,mutation_type,mutation_rate,func_mutation)
  mutation_type = "shuffle" or mutation_type
  func_mutation = nil or func_mutation
  mutation_rate = 0.50 or mutation_rate
  func_mutation = func_mutation or function(x) return math.cos(x) end
  local weights = {}
  
  --Shuffle mutation
  if mutation_type == "shuffle" then
    -- get weights
    for i=1,#neural.weights do
      weights[i] = matrix.toarray(neural.weights[i])
    end
    
    --Mutate
    for i=1,math.ceil(#weights*mutation_rate) do
      weights[i] = shuffle(weights[i])
    end
    
    -- insert mutation
    local i = 1
    while i<#neural.neurons do
      local weight_n = matrix.new(neural.neurons[i+1],neural.neurons[i],weights[i])
      neural.weights[i] = weight_n
      i = i+1
  end
    
  end
  
  --random mutation
  if mutation_type == "random" then
    for i=1,math.ceil(#weights*mutation_rate) do
      local r = math.random(1,#neural.weights[i])
      neural.weights[i] = matrix.mult1D(neural.weights[i],math.random()*math.random(-9,9))
    end
  end
  --Function mutation
  if mutation_type == 'function' then
    for i=1,#neural.weights*mutation_rate do
      neural.weights[i] = matrix.map(neural.weights[i],func_mutation)
    end
  end
  return neural
end
-- Crossover
function nn.crossover(neuralA,neuralB)

  local child = nn.new(neuralA.neurons)
  
  for i=1,#child.weights do
    if i%2 == 0 then
      child.weights[i] = neuralA.weights[i]
    else
      child.weights[i] = neuralB.weights[i]
    end
  end
  return child
end

return nn
